### Fase 1: Exploración y Limpieza
- **Exploración Inicial:**

In [168]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd # para el manejo de DataFrames
import numpy as np # para el manejo de arrays y operaciones matemáticas

# Visualización
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt # para crear gráficos
import seaborn as sns

# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames


import warnings
warnings.filterwarnings("ignore") # para evitar los mensajes de advertencia


In [169]:
df_vuelos = pd.read_csv("files/Customer Flight Activity.csv") # para cargar los datos de vuelo del cliente
df_hclientes= pd.read_csv("files/Customer Loyalty History.csv") # para cargar los datos del historial de lealtad del cliente

## Vamos a analizar en df_vuelos

In [170]:
df_vuelos.head() # para visualizar las primeras filas del DataFrame vuelos y tener una idea de su estructura

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0
2,100140,2017,1,6,0,6,1200,120.0,0,0
3,100214,2017,1,0,0,0,0,0.0,0,0
4,100272,2017,1,0,0,0,0,0.0,0,0


In [171]:
df_vuelos.shape # para conocer el número de filas y columnas del DataFrame vuelos

(405624, 10)

In [172]:
df_vuelos.info() # para conocer el tipo de datos de cada columna del DataFrame vuelos y detectar posibles problemas de formato
#Vemos que en este dataframe no hay valores nulos ya que todas las columnas tienen 405624 valores(que es el total de filas del df(como hemos visto en el shape) 

<class 'pandas.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


A continuación,vamos a revisar si encontramos algún dato que debamos cambiarle el tipo:

- Veo que sería interesante cambiar a tipo fecha las columnas "Year" y "Month"
- Eliminar espacios de los nombres de las columnas y poner "-" (buenas prácticas)
- He revisado y es correcto que el tipo de la columna Points Acumulated sea tipo float ya que hay decimales

In [173]:
#Vamos a ver si hay valores duplicados en el DataFrame vuelos
df_vuelos.duplicated().sum() # Vemos que hay 1864 filas duplicadas que eliminaremos antes de hacer la unión de los dataframes para evitar que se dupliquen datos al hacer la unión

np.int64(1864)

In [174]:
# Eliminamos las filas duplicadas del DataFrame vuelos
df_vuelos = df_vuelos.drop_duplicates() 

In [175]:
# Verificamos que se han eliminado las filas duplicadas
df_vuelos.duplicated().sum() # Vemos que efectivamente ya no hay filas duplicadas

np.int64(0)

In [176]:
#Vamos a comprobar que tal y como hemos visto en el info, no hay valores nulos
df_vuelos.isna().sum()/df_vuelos.shape[0]*100 #lo ponemos en porcentaje para verlos en valores relativos

Loyalty Number                 0.0
Year                           0.0
Month                          0.0
Flights Booked                 0.0
Flights with Companions        0.0
Total Flights                  0.0
Distance                       0.0
Points Accumulated             0.0
Points Redeemed                0.0
Dollar Cost Points Redeemed    0.0
dtype: float64

In [177]:
df_vuelos.columns

Index(['Loyalty Number', 'Year', 'Month', 'Flights Booked',
       'Flights with Companions', 'Total Flights', 'Distance',
       'Points Accumulated', 'Points Redeemed', 'Dollar Cost Points Redeemed'],
      dtype='str')

In [178]:
#Vamos a comprobar que la columna Loyalty Number a pesar de ser un identificador único. en el df_vuloes SI tiene valores duplicados
df_vuelos["Loyalty Number"].duplicated().sum()

#Un mismo cliente puede tener actividad en varios meses, por eso su Loyalty Number aparece varias veces (es lógico por lo que no debemos eliminar los duplicados)
##He revisado y las demás columnas es normal que tengan valores duplicados (Los años se repiten, los puntos pueden coincidir, etc)

np.int64(387023)

In [ ]:
df_vuelos.describe().round(2) # para obtener estadísticas descriptivas de las columnas numéricas del DataFrame vuelos y tener una idea de su distribución y posibles valores atípicos

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,403760.0,549875.38,258961.51,100018.0,326699.0,550598.0,772152.0,999986.0
Year,403760.0,2017.50,0.50,2017.0,2017.0,2018.0,2018.0,2018.0
Month,403760.0,6.50,3.45,1.0,4.0,7.0,10.0,12.0
Flights Booked,403760.0,4.13,5.23,0.0,0.0,1.0,8.0,21.0
Flights with Companions,403760.0,1.04,2.08,0.0,0.0,0.0,1.0,11.0
Total Flights,403760.0,5.17,6.53,0.0,0.0,1.0,10.0,32.0
Distance,403760.0,1214.46,1434.10,0.0,0.0,525.0,2342.0,6293.0
Points Accumulated,403760.0,124.26,146.70,0.0,0.0,53.0,240.0,676.5
Points Redeemed,403760.0,30.84,125.76,0.0,0.0,0.0,0.0,876.0
Dollar Cost Points Redeemed,403760.0,2.50,10.17,0.0,0.0,0.0,0.0,71.0


In [180]:
df_vuelos.describe(include=np.number).round(2).T # para obtener estadísticas descriptivas de las columnas numéricas
#En el info()ya hemos vist que no hay ninguna columna de tipo objectpor eso ya no es necesario analizarlo

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,403760.0,549875.38,258961.51,100018.0,326699.0,550598.0,772152.0,999986.0
Year,403760.0,2017.50,0.50,2017.0,2017.0,2018.0,2018.0,2018.0
Month,403760.0,6.50,3.45,1.0,4.0,7.0,10.0,12.0
Flights Booked,403760.0,4.13,5.23,0.0,0.0,1.0,8.0,21.0
Flights with Companions,403760.0,1.04,2.08,0.0,0.0,0.0,1.0,11.0
Total Flights,403760.0,5.17,6.53,0.0,0.0,1.0,10.0,32.0
Distance,403760.0,1214.46,1434.10,0.0,0.0,525.0,2342.0,6293.0
Points Accumulated,403760.0,124.26,146.70,0.0,0.0,53.0,240.0,676.5
Points Redeemed,403760.0,30.84,125.76,0.0,0.0,0.0,0.0,876.0
Dollar Cost Points Redeemed,403760.0,2.50,10.17,0.0,0.0,0.0,0.0,71.0


## Vamos a analizar en df_hclientes:

#Realizamos los mismos pasos que ya hemos realizado en el df_vuelos

In [181]:
df_hclientes.shape

(16737, 16)

In [182]:
df_hclientes.columns

Index(['Loyalty Number', 'Country', 'Province', 'City', 'Postal Code',
       'Gender', 'Education', 'Salary', 'Marital Status', 'Loyalty Card',
       'CLV', 'Enrollment Type', 'Enrollment Year', 'Enrollment Month',
       'Cancellation Year', 'Cancellation Month'],
      dtype='str')

In [183]:
df_hclientes.info()
#Vemos a simple vista que hay 3 columnas que tienen valores nulos (Cancellatio year, Cancellation month y salary)
#Más adelante analaizaremos los valores nulos y veremos que hacemos con ellos (Lo trataremos en la fase 2 Limpieza de datos) 

<class 'pandas.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  str    
 2   Province            16737 non-null  str    
 3   City                16737 non-null  str    
 4   Postal Code         16737 non-null  str    
 5   Gender              16737 non-null  str    
 6   Education           16737 non-null  str    
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  str    
 9   Loyalty Card        16737 non-null  str    
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  str    
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: float64(4), 

In [184]:
#Las siguientes columnas sería interesante cambiarle el tipo de dato una vez gestionados los nulos a fecha: Cancellation year, Cancellation month.
# Cambiar también el tipo de dato de Enrollment Year y Enrollment Month a fecha

In [185]:
df_hclientes.duplicated().sum()
#No hay filas duplicadas 

np.int64(0)

In [186]:
df_hclientes["Loyalty Number"].duplicated().sum()

#Vemos que en este dataframe el Loyalty number no está duplicado porque cada fila representa a un cliente diferente

np.int64(0)

In [187]:
round(df_hclientes.isna().sum()/df_hclientes.shape[0]*100,2)#lo ponemos en porcentaje para verlos en valores relativos 
#Vemos que tal y como habíamos visto en el info, las columnas con valores nulos son Cancellation year, Cancellation month y Salary.
# En la fase1.2-Limpieza de datos
#los gestionaremos y veremos si los imputamos, si elimnamos columnas porque no nos aportan información relevante...

Loyalty Number         0.00
Country                0.00
Province               0.00
City                   0.00
Postal Code            0.00
Gender                 0.00
Education              0.00
Salary                25.32
Marital Status         0.00
Loyalty Card           0.00
CLV                    0.00
Enrollment Type        0.00
Enrollment Year        0.00
Enrollment Month       0.00
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64

In [188]:
df_hclientes.describe().round(2).T
#Aquí,vemos hay valores negativos en la columna Salary y no tiene sentido,habrá que ver que error es.

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,16737.0,549735.88,258912.13,100018.00,326603.00,550434.00,772019.00,999986.00
Salary,12499.0,79245.61,35008.30,-58486.00,59246.50,73455.00,88517.50,407228.00
CLV,16737.0,7988.90,6860.98,1898.01,3980.84,5780.18,8940.58,83325.38
Enrollment Year,16737.0,2015.25,1.98,2012.00,2014.00,2015.00,2017.00,2018.00
Enrollment Month,16737.0,6.67,3.40,1.00,4.00,7.00,10.00,12.00
Cancellation Year,2067.0,2016.50,1.38,2013.00,2016.00,2017.00,2018.00,2018.00
Cancellation Month,2067.0,6.96,3.46,1.00,4.00,7.00,10.00,12.00


In [189]:
df_hclientes.describe(include=str).round(2).T
#Vemos que la columna Country solamente tiene un valor único,por lo que es constante y no nos aporta información relevante
#Seguramente es candidata a ser eliminada (lo revisaremos en la fase 1.2 Limpieza de datos)

,count,unique,top,freq
Country,16737,1,Canada,16737
Province,16737,11,Ontario,5404
City,16737,29,Toronto,3351
Postal Code,16737,55,V6E 3D9,911
Gender,16737,2,Female,8410
Education,16737,5,Bachelor,10475
Marital Status,16737,3,Married,9735
Loyalty Card,16737,3,Star,7637
Enrollment Type,16737,2,Standard,15766


In [190]:
df_hclientes["Country"].value_counts() # para confirmar que la columna Country tiene un único valor y es constante


Country
Canada    16737
Name: count, dtype: int64

In [191]:
df_hclientes["Country"].nunique()

1

Vamos a revisar los valores de las columnas que tienen pocos valores únicos para ver que contienen (ver cuales son los valores)

In [192]:
df_hclientes["Loyalty Card"].value_counts()

Loyalty Card
Star      7637
Nova      5671
Aurora    3429
Name: count, dtype: int64

In [193]:
df_hclientes["Enrollment Type"].value_counts()


Enrollment Type
Standard          15766
2018 Promotion      971
Name: count, dtype: int64

In [194]:
df_hclientes["Marital Status"].value_counts()

Marital Status
Married     9735
Single      4484
Divorced    2518
Name: count, dtype: int64

In [195]:
df_hclientes["Education"].value_counts()

Education
Bachelor                10475
College                  4238
High School or Below      782
Doctor                    734
Master                    508
Name: count, dtype: int64

## Vamos a hacer la unión de los dos dataframes anteriores

In [196]:
#Para unir los dos dataframes, se va a usar la columna Loyalty Number que es el identificador único de cada cliente y que aparece en ambos dataframes. 
#La unión la realizaré usando un left merge ya que quiero conservar la información de los clientes que aún no han comprado ni reservado ningún vuelo,


In [197]:

df_final = pd.merge(df_hclientes, df_vuelos, on="Loyalty Number", how="left") 
  #El dataframe izquierdo es el que quieres conservar completo, aunque no haya coincidencias en el otro dataframe.
  # Si pusiéramos en el left el otro dataframe, solo se conservarían los clientes que tienen actividad.

In [198]:
df_final.head(3)
#Revisamos que la unión se ha realizado correctamente 


,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN,2017,1,0,0,0,0,0.0,0,0
1,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN,2017,2,3,0,3,2823,282.0,0,0
2,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN,2017,3,0,0,0,0,0.0,0,0


In [199]:
df_final.columns    

Index(['Loyalty Number', 'Country', 'Province', 'City', 'Postal Code',
       'Gender', 'Education', 'Salary', 'Marital Status', 'Loyalty Card',
       'CLV', 'Enrollment Type', 'Enrollment Year', 'Enrollment Month',
       'Cancellation Year', 'Cancellation Month', 'Year', 'Month',
       'Flights Booked', 'Flights with Companions', 'Total Flights',
       'Distance', 'Points Accumulated', 'Points Redeemed',
       'Dollar Cost Points Redeemed'],
      dtype='str')

In [200]:
df_final.duplicated().sum()


np.int64(0)

In [201]:
df_final.shape # revisamos el número de filas y columnas del df_final

(403760, 25)

In [202]:
df_hclientes.shape

(16737, 16)

In [203]:
df_vuelos.shape

(403760, 10)

In [204]:
(df_final.isna().sum()/df_final.shape[0] *100).round(2)
#Revisamos que no hay más nulos que antes por error debido a la unión.Continuamos teniendo las columnas Salary 25%nulos,Cancellation Year 87.70%nulos y Cancellation Month 87.70%nulos

Loyalty Number                  0.00
Country                         0.00
Province                        0.00
City                            0.00
Postal Code                     0.00
Gender                          0.00
Education                       0.00
Salary                         25.33
Marital Status                  0.00
Loyalty Card                    0.00
CLV                             0.00
Enrollment Type                 0.00
Enrollment Year                 0.00
Enrollment Month                0.00
Cancellation Year              87.70
Cancellation Month             87.70
Year                            0.00
Month                           0.00
Flights Booked                  0.00
Flights with Companions         0.00
Total Flights                   0.00
Distance                        0.00
Points Accumulated              0.00
Points Redeemed                 0.00
Dollar Cost Points Redeemed     0.00
dtype: float64

- **Limpieza de Datos:**

In [205]:
#Gestionamos los valores nulos:
(df_final.isna().sum()/df_final.shape[0] *100).round(2).sort_values(ascending=False)

Cancellation Month             87.70
Cancellation Year              87.70
Salary                         25.33
Loyalty Number                  0.00
Country                         0.00
Postal Code                     0.00
Gender                          0.00
City                            0.00
Province                        0.00
Marital Status                  0.00
Education                       0.00
Loyalty Card                    0.00
CLV                             0.00
Enrollment Year                 0.00
Enrollment Type                 0.00
Enrollment Month                0.00
Year                            0.00
Month                           0.00
Flights Booked                  0.00
Flights with Companions         0.00
Total Flights                   0.00
Distance                        0.00
Points Accumulated              0.00
Points Redeemed                 0.00
Dollar Cost Points Redeemed     0.00
dtype: float64

In [206]:
#Vamos a explorar los nulos de la columna salario y buscar patrones (si tienen que ver con otras variables)
df_final[df_final['Salary'].isna()]

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
24,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN,2017,1,9,1,10,1860,186.0,0,0
25,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN,2017,2,9,0,9,2043,204.0,0,0
26,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN,2017,3,14,0,14,2002,200.0,0,0
27,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN,2017,4,0,0,0,0,0.0,0,0
28,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN,2017,5,16,0,16,3056,305.0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
403707,776187,Canada,British Columbia,Vancouver,V5R 1W3,Male,College,NaN,Single,Star,74228.52,Standard,2014,3,NaN,NaN,2018,7,4,4,8,3640,364.0,0,0
403708,776187,Canada,British Columbia,Vancouver,V5R 1W3,Male,College,NaN,Single,Star,74228.52,Standard,2014,3,NaN,NaN,2018,8,6,0,6,1890,189.0,0,0
403709,776187,Canada,British Columbia,Vancouver,V5R 1W3,Male,College,NaN,Single,Star,74228.52,Standard,2014,3,NaN,NaN,2018,9,0,0,0,0,0.0,0,0
403710,776187,Canada,British Columbia,Vancouver,V5R 1W3,Male,College,NaN,Single,Star,74228.52,Standard,2014,3,NaN,NaN,2018,11,11,0,11,2552,255.0,0,0


In [207]:
df_final[df_final['Salary'].isna()].sample(10)
#Después de haber ejecutado varios .sample, he detectado que todos los clientes que tienen salario nulo,Education es "College".(La columna Country ses siempre Canada.Lo la tengo en cuenta ya que ya la he marcado para eliminir)

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
381677,751779,Canada,Quebec,Hull,J8Y 3Z5,Male,College,NaN,Married,Star,11885.97,Standard,2014,9,NaN,NaN,2018,1,8,0,8,672,67.0,415,34
166550,546163,Canada,British Columbia,West Vancouver,V6V 8Z3,Female,College,NaN,Married,Nova,3324.55,2018 Promotion,2018,3,NaN,NaN,2017,10,0,0,0,0,0.0,0,0
180800,829162,Canada,Quebec,Montreal,H4G 3T4,Male,College,NaN,Single,Nova,7091.74,Standard,2014,1,NaN,NaN,2017,12,16,4,20,1480,148.0,0,0
355175,853621,Canada,British Columbia,Whistler,V6T 1Y8,Female,College,NaN,Single,Star,7664.81,Standard,2018,7,NaN,NaN,2018,5,0,0,0,0,0.0,0,0
241763,807297,Canada,Quebec,Hull,J8Y 3Z5,Male,College,NaN,Single,Star,2239.65,Standard,2016,4,NaN,NaN,2018,7,3,3,6,2430,243.0,0,0
15083,784838,Canada,Nova Scotia,Halifax,B3J 9S2,Male,College,NaN,Single,Star,4858.07,Standard,2015,8,NaN,NaN,2018,5,0,0,0,0,0.0,0,0
188281,566550,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Nova,7707.81,Standard,2018,8,NaN,NaN,2018,11,4,0,4,832,83.0,0,0
986,127400,Canada,British Columbia,Whistler,V6T 1Y8,Female,College,NaN,Single,Star,3914.72,Standard,2014,6,NaN,NaN,2017,7,11,0,11,4323,432.0,0,0
169705,997710,Canada,Ontario,Toronto,M8Y 4K8,Male,College,NaN,Divorced,Nova,5841.53,Standard,2013,1,2017.0,10.0,2017,6,2,0,2,1308,130.0,0,0
312700,327470,Canada,Ontario,Toronto,P2T 6G3,Male,College,NaN,Single,Star,4655.39,Standard,2013,6,NaN,NaN,2018,1,0,0,0,0,0.0,0,0


In [208]:
df_final[df_final['Salary'].isna()]['Education'] 

24        College
25        College
26        College
27        College
28        College
           ...   
403707    College
403708    College
403709    College
403710    College
403711    College
Name: Education, Length: 102260, dtype: str

In [209]:
df_final['Salary'].isna()

0         False
1         False
2         False
3         False
4         False
          ...  
403755    False
403756    False
403757    False
403758    False
403759    False
Name: Salary, Length: 403760, dtype: bool

In [210]:
df_final['Salary'].isna().groupby(df_final['Education']).sum()

#Vemos que todos los nulos de la variable salario corresponden a los clientes que su ivel educativo es College.


Education
Bachelor                     0
College                 102260
Doctor                       0
High School or Below         0
Master                       0
Name: Salary, dtype: int64

In [211]:
df_final.groupby('Education')['Salary'].mean().sort_values(ascending=False)
#Aquí estamos viendo como cambia el salario medio según el nivel educativo
#

Education
Doctor                  178403.675202
Master                  103734.523724
Bachelor                 72473.471154
High School or Below     61019.926143
College                           NaN
Name: Salary, dtype: float64

In [212]:
#Vemos que como college no tiene ningún valor,(todos son nulos),no podemos valorar si la imputacón debería ser por la media de college,por la mediana....
#Por lógica, el salario de los clientes con nivel educativo --> college debería estar entre Bachellor y High School or Bellow
#Para poder limpiar los datos y no imputar los nulos como "Desconocido" 
# y viendo que tiene sentido que el salario de los que tiene un nivel educativo "College" esté comprendido entre las otras dos categorías ya mencionadas
#Los nulos serán imputados por la media de las columnas Bachellor y High School or Bellow




In [213]:
# Filtramos solo Bachelor y High School or Below y calculamos la media
media_college = df_final[df_final['Education'].isin(['Bachelor', 'High School or Below'])]['Salary'].mean()
print(media_college)


71675.46685231433


In [214]:
#Procedemos a imputar los nulos de la columna Salary por la media que hemos calculado para college

df_final.loc[df_final['Education'] == 'College', 'Salary'] = media_college




In [215]:
df_final.groupby('Education')['Salary'].mean().sort_values(ascending=False)
#Comprobams que ya han sido imputados los nulos con la media

Education
Doctor                  178403.675202
Master                  103734.523724
Bachelor                 72473.471154
College                  71675.466852
High School or Below     61019.926143
Name: Salary, dtype: float64

In [216]:
#Ahora vamos a analizar los nulos de las columnas Cancellation Year y Cancellation Month para ver si tienen algún patrón y cómo gestionarlos.
df_final[df_final['Cancellation Year'].isna()].sample(10)

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
186851,359138,Canada,Ontario,Toronto,M8Y 4K8,Male,Bachelor,75700.000000,Married,Nova,7596.29,Standard,2018,12,NaN,NaN,2017,9,0,0,0,0,0.0,0,0
21910,235629,Canada,New Brunswick,Fredericton,E3B 2H2,Male,Master,124308.000000,Divorced,Aurora,5140.39,Standard,2018,6,NaN,NaN,2017,8,0,0,0,0,0.0,0,0
286238,568594,Canada,British Columbia,Kelowna,V09 2E9,Female,Bachelor,90200.000000,Divorced,Star,2772.44,Standard,2012,12,NaN,NaN,2017,7,0,0,0,0,0.0,0,0
204985,170566,Canada,Quebec,Montreal,H4G 3T4,Male,Bachelor,71269.000000,Married,Nova,9165.72,Standard,2017,5,NaN,NaN,2017,1,0,0,0,0,0.0,0,0
245885,927975,Canada,British Columbia,Vancouver,V5R 1W3,Female,Bachelor,69516.000000,Single,Star,2345.26,Standard,2016,1,NaN,NaN,2018,5,0,0,0,0,0.0,0,0
117864,563094,Canada,British Columbia,Vancouver,V6E 3D9,Female,College,71675.466852,Single,Star,11003.54,2018 Promotion,2018,4,NaN,NaN,2018,4,0,0,0,0,0.0,0,0
264718,312545,Canada,Quebec,Tremblant,H5Y 2S9,Female,Master,97596.000000,Divorced,Star,2538.63,Standard,2017,10,NaN,NaN,2017,6,0,0,0,0,0.0,0,0
185478,797432,Canada,Ontario,Toronto,M8Y 4K8,Female,Bachelor,60968.000000,Married,Aurora,5811.92,2018 Promotion,2018,2,NaN,NaN,2017,4,0,0,0,0,0.0,0,0
329265,734952,Canada,Ontario,Sudbury,M5V 1G5,Female,Bachelor,100613.000000,Married,Star,5386.97,Standard,2017,11,NaN,NaN,2017,3,0,0,0,0,0.0,0,0
254813,727507,Canada,British Columbia,Vancouver,V5R 1W3,Male,College,71675.466852,Single,Star,2452.36,Standard,2017,6,NaN,NaN,2017,7,0,0,0,0,0.0,0,0


In [217]:
df_final.groupby('Cancellation Year',dropna=False)['Enrollment Type'].value_counts()

Cancellation Year  Enrollment Type
2013.0             Standard             1032
2014.0             Standard             4344
2015.0             Standard             6360
2016.0             Standard            10248
2017.0             Standard            12147
2018.0             Standard            12752
                   2018 Promotion       2767
NaN                Standard           333536
                   2018 Promotion      20574
Name: count, dtype: int64

In [218]:
# Vemos que hay nulos en más de un enrollment type, por lo que no parece haber un patrón claro 
df_final[df_final['Cancellation Year'].isna()]['Enrollment Type'].value_counts()



Enrollment Type
Standard          333536
2018 Promotion     20574
Name: count, dtype: int64

In [219]:
df_final[df_final['Cancellation Month'].isna()]['Enrollment Type'].value_counts()
# Vemos que hay nulos en más de un enrollment type, por lo que no parece haber un patrón claro  y además reconfirmamos que los nulos de Cancellation Year y Cancellation Month coinciden 

Enrollment Type
Standard          333536
2018 Promotion     20574
Name: count, dtype: int64

In [220]:
df_final['Cancellation Year'].unique()

array([  nan, 2018., 2015., 2017., 2014., 2016., 2013.])

In [221]:
#Hipótesis: los nulos de Cancellation Year y Cancellation Month podrían indicar que el cliente no ha cancelado ningún vuelo:


# Agrupamos por el año de cancelación incluyendo los nulos
df_final.groupby('Cancellation Year', dropna=False)['Flights Booked'].sum()
#Los nulos Son los clientes que NO han cancelado. Han realizado más de 1.6 millones de vuelos.Se ve claramente que los nulos en estas dos columnas son los clientes activos(por eso no han cancelado)
#Si los necesitara en un futuro, los podría imputar como "Cliente activo".
#En este caso, los nulos en estas dos columnas son "NULOS CON SIGNIFICADO"


Cancellation Year
2013.0          0
2014.0          0
2015.0          0
2016.0          0
2017.0      17247
2018.0      44059
NaN       1607858
Name: Flights Booked, dtype: int64

In [222]:
#Habíamos vito que en Salario, había valores en negativo.Vamos a analizarlos 
df_final[df_final['Salary'] < 0].sample(10)

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
403727,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN,2018,4,5,4,9,2970,320.76,0,0
247330,790475,Canada,Ontario,Trenton,K8V 4B2,Female,Bachelor,-34079.0,Married,Nova,12913.50,2018 Promotion,2018,2,NaN,NaN,2018,9,5,5,10,990,99.00,575,46
345890,239955,Canada,Quebec,Hull,J8Y 3Z5,Female,Bachelor,-47310.0,Married,Nova,6366.23,2018 Promotion,2018,3,NaN,NaN,2018,4,6,5,11,1903,237.50,0,0
26156,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2017,1,0,0,0,0,0.00,0,0
345880,239955,Canada,Quebec,Hull,J8Y 3Z5,Female,Bachelor,-47310.0,Married,Nova,6366.23,2018 Promotion,2018,3,NaN,NaN,2017,5,0,0,0,0,0.00,0,0
159055,430398,Canada,Newfoundland,St. John's,A1C 6H9,Male,Bachelor,-17534.0,Married,Nova,49423.80,2018 Promotion,2018,3,NaN,NaN,2017,1,0,0,0,0,0.00,0,0
281084,366599,Canada,Ontario,Toronto,M1R 4K3,Female,Bachelor,-9081.0,Married,Star,6915.73,2018 Promotion,2018,4,NaN,NaN,2017,10,8,0,8,1096,109.00,0,0
26170,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2018,4,5,0,5,985,105.84,0,0
403718,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN,2017,7,6,0,6,4518,451.00,0,0
304219,436087,Canada,Quebec,Montreal,H2T 9K8,Male,Bachelor,-46470.0,Married,Star,4786.89,2018 Promotion,2018,4,2018.0,8.0,2018,9,0,0,0,0,0.00,0,0


In [223]:
df_final[df_final['Salary'] > 0].sample(10)

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
375412,816608,Canada,Quebec,Montreal,H4G 3T4,Female,Bachelor,59169.000000,Married,Star,9816.53,Standard,2015,9,NaN,NaN,2017,1,3,3,6,2394,239.0,0,0
143175,295906,Canada,British Columbia,Vancouver,V6E 3D9,Male,Bachelor,70658.000000,Married,Nova,4530.39,Standard,2013,4,NaN,NaN,2017,9,0,0,0,0,0.0,0,0
193956,822665,Canada,Ontario,Trenton,K8V 4B2,Female,College,71675.466852,Married,Nova,8181.72,Standard,2018,12,NaN,NaN,2017,11,1,1,2,1558,155.0,0,0
70351,593651,Canada,British Columbia,Vancouver,V5R 1W3,Female,Bachelor,47297.000000,Married,Aurora,9320.28,Standard,2014,12,NaN,NaN,2017,2,12,0,12,1884,188.0,0,0
399083,423313,Canada,Ontario,Trenton,K8V 4B2,Male,Master,81765.000000,Single,Star,25464.82,Standard,2015,8,2016.0,4.0,2017,4,0,0,0,0,0.0,0,0
46828,822727,Canada,Ontario,Trenton,K8V 4B2,Female,College,71675.466852,Single,Aurora,7086.18,Standard,2015,11,2018.0,6.0,2018,8,0,0,0,0,0.0,0,0
2504,890753,Canada,Quebec,Montreal,H2Y 2W2,Male,College,71675.466852,Married,Star,4025.10,Standard,2017,9,NaN,NaN,2018,1,0,0,0,0,0.0,0,0
200564,823816,Canada,British Columbia,Vancouver,V1E 4R6,Male,Bachelor,103449.000000,Divorced,Nova,8692.64,Standard,2016,8,NaN,NaN,2018,11,0,0,0,0,0.0,0,0
7683,766927,Canada,Ontario,Toronto,P1J 8T7,Male,Bachelor,71936.000000,Divorced,Star,4457.94,Standard,2016,4,NaN,NaN,2018,7,3,3,6,4752,475.0,771,62
237004,366687,Canada,British Columbia,Kelowna,V09 2E9,Female,Bachelor,48021.000000,Single,Nova,46611.87,Standard,2018,7,NaN,NaN,2018,10,3,0,3,1149,114.0,0,0


In [224]:
df_final[['Salary']].describe().T.round(2)
#Vemos que el salario minimo está en negativo, pero que si estuviera en positivo, no estaría desequilibrado

,count,mean,std,min,25%,50%,75%,max
Salary,403760.0,77338.01,30410.19,-58486.0,63901.0,71675.47,82940.0,407228.0


In [225]:
df_final[df_final['Loyalty Number'] == 491242]
#Analizamos un cliente con salario negativo y vemos que todos sus registros son negativos


,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
114057,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,1,0,0,0,0,0.00,0,0
114058,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,2,0,0,0,0,0.00,0,0
114059,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,3,0,0,0,0,0.00,0,0
114060,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,4,0,0,0,0,0.00,0,0
114061,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,5,0,0,0,0,0.00,0,0
114062,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,6,0,0,0,0,0.00,0,0
114063,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,7,0,0,0,0,0.00,0,0
114064,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,8,0,0,0,0,0.00,0,0
114065,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,9,0,0,0,0,0.00,0,0
114066,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN,2017,10,0,0,0,0,0.00,0,0


In [226]:
df_final[df_final['Loyalty Number'] == 347013]
#Vemos lo mismo


,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
346548,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,1,0,0,0,0,0.0,0,0
346549,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,2,1,1,2,702,70.0,0,0
346550,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,3,0,0,0,0,0.0,0,0
346551,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,11,8,1,9,2538,253.0,317,26
346552,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,4,12,0,12,1308,130.0,0,0
346553,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,5,9,0,9,2529,252.0,0,0
346554,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,6,0,0,0,0,0.0,0,0
346555,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,7,17,6,23,1817,181.0,0,0
346556,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,8,2,0,2,268,26.0,0,0
346557,347013,Canada,Quebec,Quebec City,G1B 3L5,Male,Bachelor,-39503.0,Married,Nova,6446.71,2018 Promotion,2018,3,NaN,NaN,2017,9,0,0,0,0,0.0,0,0


In [227]:
salario_negativo=df_final[df_final['Salary'] < 0]


In [228]:
df_final[df_final['Salary'] < 0]

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
26156,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2017,1,0,0,0,0,0.0,0,0
26157,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2017,2,0,0,0,0,0.0,0,0
26158,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2017,3,0,0,0,0,0.0,0,0
26159,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2017,4,0,0,0,0,0.0,0,0
26160,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN,2017,5,0,0,0,0,0.0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
403731,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN,2018,8,0,0,0,0,0.0,0,0
403732,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN,2018,9,8,0,8,2120,212.0,0,0
403733,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN,2018,10,0,0,0,0,0.0,0,0
403734,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN,2018,11,3,3,6,822,82.0,0,0


In [229]:
negativos = df_final[df_final['Salary'] < 0]['Loyalty Number'].unique()
print(negativos)
#Son los clientes que tienen salario negativo


[542976 959977 232755 525245 603070 491242 115505 430398 152016 194065
 212128 790475 366599 436087 364596 239955 347013 729561 734647 906428]


In [237]:
#Veo que es un error de registro, ya que el mismo cliente tiene varios registros con salario negativo, por lo que no es un error puntual, sino que es un error sistemático.

df_final['Salary'] = df_final['Salary'].abs() 

In [238]:
negativos = df_final[df_final['Salary'] < 0]['Loyalty Number'].unique()
print(negativos)

#Ya no hay clientes con salario negativo

[]


In [241]:
#Vamos a eliminar la columna country que es constante y no nos aporta información relevante
df_final = df_final.drop(columns=['Country'])


In [ ]:
df_final.sample()
#Vemos que ya hemos elimindo la columna Country.


,Loyalty Number,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
203861,771353,Quebec,Montreal,H4G 3T4,Female,Bachelor,53862.0,Married,Nova,9033.25,Standard,2017,3,2017.0,11.0,2018,1,0,0,0,0,0.0,0,0


In [245]:
#Vamos a poner el nombre de las columnas en minúscula y con guiones bajos por buenas prácticas y para facilitar su uso en el futuro
df_final.columns = df_final.columns.str.lower().str.replace(' ', '_')

In [ ]:
df_final.sample()
#comprobamos que ya hemos cambiado el nombre de las columnas a minúscula y con guiones bajos

,loyalty_number,province,city,postal_code,gender,education,salary,marital_status,loyalty_card,clv,enrollment_type,enrollment_year,enrollment_month,cancellation_year,cancellation_month,year,month,flights_booked,flights_with_companions,total_flights,distance,points_accumulated,points_redeemed,dollar_cost_points_redeemed
160496,478127,Ontario,Toronto,P1J 8T7,Female,Bachelor,80516.0,Married,Nova,5389.5,Standard,2015,11,NaN,NaN,2017,2,3,3,6,2130,213.0,0,0


In [249]:
df_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 403760 entries, 0 to 403759
Data columns (total 24 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   loyalty_number               403760 non-null  int64  
 1   province                     403760 non-null  str    
 2   city                         403760 non-null  str    
 3   postal_code                  403760 non-null  str    
 4   gender                       403760 non-null  str    
 5   education                    403760 non-null  str    
 6   salary                       403760 non-null  float64
 7   marital_status               403760 non-null  str    
 8   loyalty_card                 403760 non-null  str    
 9   clv                          403760 non-null  float64
 10  enrollment_type              403760 non-null  str    
 11  enrollment_year              403760 non-null  int64  
 12  enrollment_month             403760 non-null  int64  
 13  cancellati

In [256]:
# Comparamos antes de guardar el dataframe en un csv que los float siguientes si habíamos visto en un head, que tenían decimales.
#De esta forma, nos aseguramos de que está correcto el formato.

tiene_decimales = (df_final['clv'] % 1 != 0).any()
print(tiene_decimales)


True


In [257]:


tiene_decimales = (df_final['salary'] % 1 != 0).any()

print(tiene_decimales)

True


In [258]:


tiene_decimales = (df_final['points_accumulated'] % 1 != 0).any()
print(tiene_decimales)

True


In [ ]:
#Ahora que ya tenemos un df unido con los datos de los dos csvs, los datos limpios y preparados vamos a ver si se ve algún valor atípico a simple vista
#Vamos a analizar si con el describe de las variables numéricas vemos algún valor atípico:
df_final.describe().round(2).T


,count,mean,std,min,25%,50%,75%,max
loyalty_number,403760.0,549875.38,258961.51,100018.00,326699.00,550598.00,772152.00,999986.00
salary,403760.0,77422.51,30194.42,9081.00,63901.00,71675.47,82940.00,407228.00
clv,403760.0,7990.86,6863.32,1898.01,3981.78,5776.34,8937.12,83325.38
enrollment_year,403760.0,2015.25,1.98,2012.00,2014.00,2015.00,2017.00,2018.00
enrollment_month,403760.0,6.67,3.40,1.00,4.00,7.00,10.00,12.00
cancellation_year,49650.0,2016.50,1.38,2013.00,2016.00,2017.00,2018.00,2018.00
cancellation_month,49650.0,6.96,3.45,1.00,4.00,7.00,10.00,12.00
year,403760.0,2017.50,0.50,2017.00,2017.00,2018.00,2018.00,2018.00
month,403760.0,6.50,3.45,1.00,4.00,7.00,10.00,12.00
flights_booked,403760.0,4.13,5.23,0.00,0.00,1.00,8.00,21.00


Posibles outliers:(valor atípico):

- SALARIO: Vemos que el salario minimo y maximo es muy diferente (9k vs 407k ).Por otro lado también vemos que está muy alejado de la media que son 77k.Además la mediana es 71k.También vemos que el 75% de los clientes cobran menos de 82k

- CLV: El valor mínimo y el valor maximo están muy alejados:min 1,8k y max 83k.Aquí vemos que hay un cliente estrella pero no se acerca a lo habitual.También lo vemos en que el 75 % de los clientes gasta menos de 9k.La media está en 8k y la mediana en 5.7k.La media está empujada a la alza por los outliers.
- TOTAL FLIGHTS:Vemos que 75% de los clients realiza menos de 10 viajes, sin embargo el maximos es 32 viajes.

-DISTANCE:Vemos que la
-POINTS REDEEMED:Vemos que la media de puntos canjeados es de 30 puntos, sin embargo vemos que el maximo es 876 puntos.Una diferencia bastante grande



In [ ]:
#posibles outliers en salary, clv y points_accumulated, aunque no parecen estar tan alejados del resto de valores como para considerarlos outliers extremos.


In [ ]:
OJOOOO GUARDARLO